File: src/Atlasd/Battlenet/Protocols/Game/Messages/SID_CLANMEMBERLIST.cs
using Atlasd.Battlenet.Exceptions;
using Atlasd.Daemon;
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;

namespace Atlasd.Battlenet.Protocols.Game.Messages
{
    class SID_CLANMEMBERLIST : Message
    {
        public SID_CLANMEMBERLIST()
        {
            Id = (byte)MessageIds.SID_CLANMEMBERLIST;
            Buffer = new byte[0];
        }

        public SID_CLANMEMBERLIST(byte[] buffer)
        {
            Id = (byte)MessageIds.SID_CLANMEMBERLIST;
            Buffer = buffer;
        }

        public override bool Invoke(MessageContext context)
        {
            switch (context.Direction)
            {
                case MessageDirection.ClientToServer:
                    {
                        Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)");

                        if (context.Client.GameState.ActiveAccount == null)
                        {
                            throw new GameProtocolViolationException(context.Client, $"{MessageName(Id)} received before logon");
                        }

                        /**
                         * (UINT32) Cookie
                         */

                        if (Buffer.Length != 4)
                        {
                            throw new GameProtocolViolationException(context.Client, $"{MessageName(Id)} buffer must be 4 bytes");
                        }

                        using var m = new MemoryStream(Buffer);
                        using var r = new BinaryReader(m);

                        var cookie = r.ReadUInt32();

                        return new SID_CLANMEMBERLIST().Invoke(new MessageContext(context.Client, MessageDirection.ServerToClient, new Dictionary<string, dynamic>(){
                            { "cookie", cookie },
                        }));
                    }
                case MessageDirection.ServerToClient:
                    {
                        var cookie = (UInt32)context.Arguments["cookie"];

                        /**
                         * (UINT32) Cookie
                         *  (UINT8) Member Count
                         *
                         * For each member:
                         *   (STRING) Username
                         *    (UINT8) Rank
                         *    (UINT8) Online Status
                         *   (STRING) Location
                         */

                        Buffer = new byte[5];

                        using var m = new MemoryStream(Buffer);
                        using var w = new BinaryWriter(m);

                        w.Write(cookie);
                        w.Write((byte)0);

                        Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)");
                        context.Client.Send(ToByteArray(context.Client.ProtocolType));
                        return true;
                    }
            }

            return false;
        }
    }
}

File: src/AtlasV/Battlenet/Protocols/Game/Messages/SID_CLANMEMBERLIST.vb
Imports AtlasV.Battlenet.Exceptions
Imports AtlasV.Daemon
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Text

Namespace AtlasV.Battlenet.Protocols.Game.Messages
    Class SID_CLANMEMBERLIST
        Inherits Message

        Public Sub New()
            Id = CByte(MessageIds.SID_CLANMEMBERLIST)
            Buffer = Array.Empty(Of Byte)()
        End Sub

        Public Sub New(ByVal varBuffer As Byte())
            Id = CByte(MessageIds.SID_CLANMEMBERLIST)
            Buffer = varBuffer
        End Sub

        Public Overrides Function Invoke(ByVal context As MessageContext) As Boolean
            Select Case context.Direction
                Case MessageDirection.ClientToServer
                    Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)")

                    If context.Client.GameState.ActiveAccount Is Nothing Then
                        Throw New GameProtocolViolationException(context.Client, $"{MessageName(Id)} received before logon")
                    End If

                    If Buffer.Length <> 4 Then
                        Throw New GameProtocolViolationException(context.Client, $"{MessageName(Id)} buffer must be 4 bytes")
                    End If

                    Dim cookie As UInt32
                    Using m = New MemoryStream(Buffer)
                        Using r = New BinaryReader(m)
                            cookie = r.ReadUInt32()
                        End Using
                    End Using

                    Return New SID_CLANMEMBERLIST().Invoke(New MessageContext(context.Client, MessageDirection.ServerToClient, New Dictionary(Of String, Object)() From {
                        {"cookie", cookie}
                    }))
                Case MessageDirection.ServerToClient
                    Dim cookie = CUInt(context.Arguments("cookie"))
                    Buffer = New Byte(4) {}

                    Using m = New MemoryStream(Buffer)
                        Using w = New BinaryWriter(m)
                            w.Write(cookie)
                            w.Write(CByte(0))
                        End Using
                    End Using

                    Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)")
                    context.Client.Send(ToByteArray(context.Client.ProtocolType))
                    Return True
            End Select

            Return False
        End Function
    End Class
End Namespace
